node/src/chain/acceptor.rs

```diff
@@ -246,6 +252,84 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
         Ok(())
     }
 
+    /// Implements the algorithm of full revert to any of supported targets.
+    ///
+    /// This incorporates both VM state revert and Ledger state revert.
+    pub async fn try_revert(&self, target: RevertTarget) -> Result<()> {
+        let curr_height = self.get_curr_height().await;
+        let curr_iteration = self.get_curr_iteration().await;
+
+        let target_state_hash = match target {
+            RevertTarget::LastFinalizedState => {
+                info!("Revert VM to last finalized state");
+                let state_hash = self.vm.read().await.revert()?;
+
+                info!(
+                    "VM revert completed finalized_state_hash:{}",
+                    hex::encode(state_hash)
+                );
+
+                anyhow::Ok(state_hash)
+            }
+            RevertTarget::LastEpoch => panic!("not implemented"),
+        }?;
+
+        // Delete any block until we reach the target_state_hash, the
+        // VM was reverted to.
+
+        // The blockchain tip (most recent block) after reverting
+        let mut most_recent_block = Block::default();
+
+        self.db.read().await.update(|t| {
+            let mut height = curr_height;
+            while height != 0 {
+                let blk = Ledger::fetch_block_by_height(t, height)?
+                    .ok_or_else(|| anyhow::anyhow!("could not fetch block"))?;
+
+                if blk.header.state_hash == target_state_hash {
+                    most_recent_block = blk;
+                    break;
+                }
+
+                info!(
+                    "Delete block height: {} iter: {} hash: {}",
+                    blk.header.height,
+                    blk.header.iteration,
+                    hex::encode(blk.header.hash)
+                );
+
+                // Delete any rocksdb record related to this block
+                Ledger::delete_block(t, &blk)?;
+
+                // Attempt to resubmit transactions back to mempool.
+                // An error here is not considered critical.
+                for tx in blk.txs().iter() {
+                    Mempool::add_tx(t, tx).map_err(|err| {
+                        tracing::error!("failed to resubmit transactions")
+                    });
+                }
+
+                height -= 1;
+            }
+
+            Ok(())
+        })?;
+
+        if most_recent_block.header.state_hash != target_state_hash {
```

Comment:
If we revert up to the genesis block, the MRB will be the self-generated block and the state_hash will differ.
Should you explicitly disable this check if the reverted height is 0?

Reply:
I designed the algorithm of revert with assumption that we'll need to revert to a finalized state (not genesis).  Even if I add this check here it will again fail to revert to genesis state.

Reply:
In other words, these scenarios will fail:

- try_revert is called at height 1, it fails to revert to genesis state (height 0). 
- try_revert is called to any height but all blocks prior to genesis are non-finalized.


@herr-seppia If you believe these are valid edge cases, I can address them too.

Reply:
Those are real, but no need to address them